### ISSUE TITLE
Performance (FPS) Issues Related to Terrain Scatters with 1.7.1 Release (Edit: And 1.7.3)

### ISSUE THREAD

**Issue by Joker58th** [Kopernicus/Kopernicus#365]
In my career save, I noticed that my framerate drops significantly on a descent to landing on the Mun under 10km.  In doing some basic testing with clean KSP 1.7.1 install it appears that having terrain scatters turned on is the cause.  

Test Environment:
KSP 1.7.1
1920x1080
Windowed
I5-9600K
RTX2080 8GB
Launch vehicle, set orbit to Mun, 90 inclination, lowest altitude possible (@15km)
Orient to retrograde and slow to 200m/s
At 2000m AGL start burn
Perform normal landing

Mods Installed:
Kopernicus 1.7.1-5
MFI 1.2.6.0
MM 4.0.2

Chart overview: As stated above, this is a recording of a descent to landing on the Mun from 15km.  The bump you see near the end is the actual touch down.

Baseline:
Clean 1.7.1, No mods
![image](https://user-images.githubusercontent.com/32574965/59488058-769a6780-8e4c-11e9-9b85-38efead68562.png)

Kopernicus Installed:
![image](https://user-images.githubusercontent.com/32574965/59488113-9b8eda80-8e4c-11e9-9a7f-21e06a0ed453.png)

Kopernicus Installed, Terrain Scatters disabled:
![image](https://user-images.githubusercontent.com/32574965/59488242-093b0680-8e4d-11e9-9247-31cebbaed14d.png)

**Comment by rothomp3** [comment]
I have also seen this issue with my install. Thanks for the pointer to Terrain Scatters though, I can play with Kopernicus again, haha. The frame rate drop was so severe on my system that the game was completely unplayable after landing.

**Comment by MiracleMagician** [comment]
I would like to point out that breaking ground scatters and fairings also causes significant lag for me, I have 5 second per frame on JNSQ's duna, and a rocket with a big fairing drops my fps below 30. Don't know whether that fairing part is related to kopernicus though, maybe caused by other mods...

**Comment by Galileo88** [comment]
@MiracleMagician do you have the latest JNSQ? The latest version fixes the FPS issue. I don’t know about the fairing thing, but the scatter performance problem was tackled in JNSQ. Perhaps you missed an update?

**Comment by rothomp3** [comment]
I don’t have any other mods that affect graphics (whether textures, scatters, or other shaders). Turning off terrain scatters makes the game playable with Kopernicus, but I still get a huge FPS hit on the surface of most planets/moons. Strangely not on Minmus. I’m an iOS dev, but I know C#… I could try to fix it myself if I had any idea what Kopernicus could be doing to affect this.

**Comment by MiracleMagician** [comment]
> 
> 
> @MiracleMagician do you have the latest JNSQ? The latest version fixes the FPS issue. I don’t know about the fairing thing, but the scatter performance problem was tackled in JNSQ. Perhaps you missed an update?
> 
> OR you didn’t select the JNSQ_(Low/Medium/High) setting in the graphic settings on the main menu.
> 
> While there is absolutely an FPS loss with Kopernicus, I was able to offset that in JNSQ and get back to the performance one would expect.

Hmmm, that's weird, I'm using version 0.7, and my terrain settings is at JNSQ_Default.

**Comment by Sigma88** [comment]
Are you using the latest SD?
0.10.3 just introduced a performance fix

**Comment by MiracleMagician** [comment]
Of course yes, now I'll try remove SD and see whether that performance issue is still there..

**Comment by MiracleMagician** [comment]
Oof, no lag without SD...

**Comment by MiracleMagician** [comment]
60 FPS, looks like it's caused by the amount of BG scatter drawn instead of some code problems...

**Comment by Sigma88** [comment]
You can try changing the scatter settings in SD

**Comment by Starwaster** [comment]
A few things
This issue is from the ROCs which are not terrain scatter; they are PQSMods. As such ROCs aren't controlled by the scatter setting, they are controlled by a frequency field in their definition nodes which affects their distribution.
The real issue with them though is probably not their distribution. The issue is that they are spawning when they are not supposed to. For performance issues they are not supposed to spawn further away from the active vessel than 2.5km. (this might be variable; I think they are actually controlled by one of the vessel loading ranges)

They are also not supposed to spawn at all when the active vessel's surface speed is faster than 500 m/s. Again, that's for performance reasons.

With Kopernicus they are spawning at ranges farther than 2.5km (I've seen them as far away as 38km) and at speeds of over 1.2 km/s. I'm not sure what role if any Sigma Dimensions is actually playing in this.  I'd like to test it without Sigma but I'm not sure if there are any Kerbin rescale mods that do not use Sigma. (can't test with RSS because ROCs are defined per CelestialBody and there aren't any ROC configs for RSS yet)

**Comment by Starwaster** [comment]
@Galileo88 would you elaborate please about JNSQ? How did you address the FPS issue there?

**Comment by rothomp3** [comment]
@Starwaster I am not using any scaling mod, I'm using stock scale and only the Outer Planets Mod, and I'm experiencing the performance degradation.

**Comment by Joker58th** [comment]
@Starwaster Unless I'm misunderstanding you, my issue description shows that I was not using any rescale mods.  Just base game with Kopernicus and MFI.

**Comment by Starwaster** [comment]
@Joker58th  @rothomp3 Yes, it's as I suspected, it's because ROCs are being spawned when they shouldn't be. I see what Galileo did now, he cut the frequency in some of the ROC_DEFINITION nodes which will reduce how many of them are created. But it doesn't change the underlying cause of the problem.

Orbiting any body, you should never see even a single ROC regardless of scatter settings because (as above) there are range based limits and active vessel velocity limits that prevent it. Press alt-F12 and turn on the Surface Feature Finder under Breaking Ground. (has to be a save game created after you got Breaking Ground). Any ROCs will be highlighted with a red pointy thing over them. If you're descending from orbit you shouldn't see any of those red indicators until you're 2.5km away from them and travelling less than 500 m/s. If you try that now you'll see those red indicators from a lot farther away.

**Comment by Galileo88** [comment]
The biggest issue is not actually the ROC, it’s the grass.

**Comment by Starwaster** [comment]
@Galileo88 not so. I'm over Tylo right now with all scatters turned _off_ and it's taking 9 seconds for one game second to elapse. If I reload with ROCs turned off then I get normal performance. With them turned on, ROCs are loading in at an altitude of about 40km and I'm doing 6km/s. 

I'm not making this up: _They should not be loaded at all at distances that great or velocities that great. That information comes directly from the developers of the game._ It is because those limitations are somehow being bypassed that they are being allowed to load in at all. Those limitations exist to prevent ROCs from degrading performance. That information also comes directly from the developers of the game.

**Comment by Joker58th** [comment]
@Starwaster, I confirmed that my testing above was without any of the expansion packs installed so breaking ground doesn't have anything to do with this issue.  However, as you are seeing maybe there is another issue in a rescale environment.  Maybe they are related, maybe not.  From the release notes of Kopernicus 1.7.1-1, the only thing that looks scatter related was this.

![image](https://user-images.githubusercontent.com/32574965/61159949-d88dd180-a4cb-11e9-8e93-8974d85d46d9.png)

I'm not sure this is the problem but thought I would mention it.

**Comment by Starwaster** [comment]
I'll just bugger off then. 

Looking at 27942 ROCs now though... shouldn't be that many. Clearly they're being forced to load in  in a situation where the ROC manager itself wouldn't load them.

**Comment by Joker58th** [comment]
> I'll just bugger off then.

Not at all.  You might be on to another issue or like I said maybe they are related somehow.  Its still good information you provide.  I'm just trying to be as clear as possible for whenever, if ever, someone looks to correct it.

Also, I just tried a quick test in KSP 1.7.3 (no expansions) and Kopernicus 1.7.3-1 and the issue persists.

**Comment by Galileo88** [comment]
@Starwaster I never said the ROC weren't contributing to poor performance, however, I did conclude that grass on kerbin, is not rendering correctly with the stock mesh when kopernicus is installed and creating a missive performance hit. Quite a few others have tested and confirmed that as well. I have seen some reports about even just the boulders hit performance now, but I haven't seen that in my testing, so I can't say you are right or wrong.

**Comment by Galileo88** [comment]
I'm not 100% why the change was made to make the positions of each scatter known. Perhaps thats what is causing the problem. Perhaps as one mesh, the performance is better?

**Comment by StollD** [comment]
> I'm not 100% why the change was made to make the positions of each
> scatter known. Perhaps thats what is causing the problem. Perhaps as
> one mesh, the performance is better?

I *think* the performance as a single mesh is better (unless you use
something like GPU instancing on DX11 / OpenGL perhaps). Unfortunately,
Kopernicus needs to know about the positions of each object, to
implement stuff like adding components (HeatEmitter, LightEmitter - ok
that one is glitchy), or enabling / disabling PartModules if they are
in range of a scatter object. 

However, that applies only to the terrain scatters, not to the ROCs
from the DLC. I don't know why Kopernicus would prossibly force them to
be active all the time, but I can take a look. Also, as far as I
remember, the ROCs (just like terrain scatter) are coupled to the same
LOD system that loads and unloads the terrain quads. They might even be
subject to subdivision.

**Comment by Galileo88** [comment]
>Also, as far as I
remember, the ROCs (just like terrain scatter) are coupled to the same
LOD system that loads and unloads the terrain quads. They might even be
subject to subdivision.

That's why I lowered the minDistance drastically. It helped tremendously with performance too.

**Comment by lordcirth** [comment]
So, I am playing with Sigma Dimensions at 4x, 1.7.3 with Making History, Breaking Ground, and a bunch of other mods. As soon as I get close to the Mun, the framerate drops to unplayable levels. I see people here discussing settings that work around the issue. What is the simplest workaround that I can use?
Thanks!

**Comment by StollD** [comment]
Ok, I finally had some time to look into this, and this is what I found
out:

First of all it seems that this issue is caused or at least heavily
exaggerated by rescaling the system. The reason for that is, that if
you rescale the system, the planets have to subdivide themselves way
earlier than at stock scale.

In KSP, the distance at wich a certain subdivision level is reached on
the terrain is calculated using this formula: 

```
radius * minDetailDistance / 2 ^ subdivision_level
``` 

Also, ROCs are always activated if their quad reaches it's max.
subdivision level.

Now if we can compare stock Kerbin and JNSQ Kerbin:

Stock:
 * radius = 600000
 * minDetailDistance = 8
 * maxLevel = 10

This means that on stock Kerbin, a terrain quad reaches its max.
subdivision if the vessel is less than about 4km away from it.

JNSQ:
 * radius = 1600000
 * minDetailDistance = 10
 * maxLevel = 10

Using those values gives us that in JNSQ, all ROCs within 15km of the
active vessel get activated. This area is 11 times bigger than the
stock one, so you get 11x the amount of active ROCs.

JNSQ has a custom terrain preset, that reduces the minDetailDistance of
Kerbin to 6, which changes the distance value to 9km. (I didnt
calculate the new area). Still more than stock, but not as many as with
the original value.

Now, we could do a 10x system too
 * radius = 6000000
 * minDetailDistance = 8
 * maxLevel = 10

If we calculate that, we get that all ROCs within 46km of the active
vessel get loaded, which results in an area that is 100 times bigger
than stock.

So, basically, what this tells us is that this is not a Kopernicus bug.
At least I don't think so. Rather, it is a bad combination of a stock
LOD system, and planet configs that are not 100% finetuned for their
respective scale with regards to the ROC system. And I am afraid I
don't think that there is anything that can be done about this in
Kopernicus directly.

If anyone has and idea I am open to suggestions. But unless Squad
changes how ROCs are loaded, the only solution might be to tweak the
PQS detail settings in planet packs until the performance is
acceptable.

**Comment by Starwaster** [comment]
@StollD Except it's not entirely based on subdivision because there is still the speed limit to consider which is that by default no ROCs are generated (in stock) over 500 m/s, the consideration being that at that speed no ROCs the player might see at that time will be relevant since they won't be easily able to land and interact with them.

**Comment by VedaantAchuthan** [comment]
Can I mention I'm getting the same issue in OPM 1.7.1 as well? I have massive framedrops close to the ground when the scatters kick in, and my ROCs arent even active due to the fact that it's a pre-existing save.

**Comment by Starwaster** [comment]
Then you're probably having the other issue; you might want to try out this person's workaround here:
https://forum.kerbalspaceprogram.com/index.php?/topic/181547-173-1-backports-kopernicus-kittopiatech/&do=findComment&comment=3640829

(note that you shouldn't do it to Jool or the sun)

**Comment by lordcirth** [comment]
> radius * minDetailDistance / 2 ^ subdivision_level

So, will lowering minDetailDistance for scaled-up systems mitigate the performance issue?

**Comment by Starwaster** [comment]
Yes it will for sure. Don't lower it too much. JNSQ has some additional terrain quality settings where I think they're all set to 4 and I don't think you should go lower than that. I haven't gone as low as 3 but 1 and 2 are right out. You will mung the terrain badly and any vessels on the surface tend to fall through.

Given the formula @StollD cited, that probably depends on the radius but I'm too tired to do the math on it.

I also have another idea regarding the speed threshold in settings.cfg (SERENITY_ROCS_VISUAL_SPEED). I indicated earlier that I thought Kopernicus was bypassing that limitation somehow but I don't think it's that simple. I think it's not a simple case of checking against the active vessel's speed. It's checking against something else that is changing when the planet is scaled up because I tried setting the limit to 0 which if my previous theory were correct should have had no effect and ROCs should have still been there. But when I loaded up my test craft they were all gone, _just as they should have been._

I did some poking around what's actually being checked is quad.sphereRoot.targetSpeed < SERENITY_ROCS_VISUAL_SPEED. If that check fails then ROCs don't get generated.

Anyone know what that is exactly? If not I'll snoop deeper

**Comment by Starwaster** [comment]
Ok.... the 'target' here is most likely the camera transform and targetSpeed is being determined by comparing the changes in target position in local space from frame to frame. (actual InverseTransformPoint of the camera transform from the PQS transform)

**Comment by Starwaster** [comment]
ok for a 10x scale SD Kerbin, changing ROC speed to 5 keeps them from showing up until I'm coming in for final approach. Used equivalent of JNSQ_High (minDetailDistance of 6) and ROC frequency scaled to 10% - it probably should have been less to get to stock numbers (1%) but they were spread out so thin and I'm lazy and didn't want to walk that far to get to them.

So I end up with about 2000 total surface features which does drop my FPS but overall performance is very playable

so I think in addition to lowering minDetailDistance, lowering SERENITY_ROCS_VISUAL_SPEED to 5 for 10x system and... 25 for JNSQ. kind of guessing on that last once since I'm still not quite clear why the apparent speed being checked is so much lower in a scaled up system. Unless a transform itself is getting scaled somewhere.  (InverseTransformPoint is sensitive to scaling and that's what Squad is using in their code)

**Comment by Joker58th** [comment]
@StollD, I read your comments but I didn't see anything about what could be causing the FPS drop on a stock install as shown in my graphs above.  Is there any idea on that?

**Comment by Galileo88** [comment]
@Joker58th you mean a stock game without Kopernicus?

If you mean with Kopernicus, the issue that people are discussing above are still true even for a stock game. The fix would be to either wait for the next update, or try the cfg that is posted in the Kopernicus thread on the forum.

**Comment by Joker58th** [comment]
@Galileo88, yes, the stock game with Kopernicus as specified in my description of the issue (which has no mention of any rescale).  From what I'm understanding of @StollD's response so far, he is saying there is no Kopernicus bug as it pertains to a rescale setups.  Am I reading that wrong?  In any case, I'm playing stock and would be using Kopernicus only for KSPRC textures.  Apologies if I'm mis-understanding here.

**Comment by StollD** [comment]
> From what I'm understanding of @StollD's response so far, he is saying
> there is no Kopernicus bug as it pertains to a rescale setups.

What I meant was that the bug that Starwaster and co. described, which 
could be mitigated by disabling the ROCs, didn't seem to be caused by 
Kopernicus, but by using it to rescale the system and the stock code not 
reacting to that in an optimized way.

There are other reasons why your framerate could drop, most notably the 
Ocean bug which has been discussed on the last pages of the Kopernicus 
thread as Galileo mentioned. You can find a config that might fix your 
performance problems there. This is a bug and will be fixed by the next 
release.

**Comment by Joker58th** [comment]
@StollD Thanks for your response.  The way I was reading about this patch was that again, it was for rescale setups.  So I just did another quick test in stock 1.7.3, Kopernicus 1.7.3-1 and MFI 1.2.6.0 and added the patch from the forum.  



![image](https://user-images.githubusercontent.com/32574965/62068219-a2f32300-b203-11e9-98d0-5df1412c96ab.png)

The top graph is with Kopernicus/MFI.  The bottom graph is clean 1.7.3 no mods.  Does that look right to you?

![image](https://user-images.githubusercontent.com/32574965/62067582-1005b900-b202-11e9-829e-0c26fceaee9a.png)


![image](https://user-images.githubusercontent.com/32574965/62067657-39264980-b202-11e9-9a92-546d8a9116c6.png)

![image](https://user-images.githubusercontent.com/32574965/62067690-4fcca080-b202-11e9-9af8-57b13b460362.png)

**Comment by Joker58th** [comment]
Here's one more with Kopernicus installed and Terrain Scatters disabled and without the patch.  Seems reasonably close to stock performance.

![image](https://user-images.githubusercontent.com/32574965/62069870-40038b00-b207-11e9-9f9f-d41b8d051fe4.png)

![image](https://user-images.githubusercontent.com/32574965/62069919-66292b00-b207-11e9-93c3-1bbd369ce982.png)

**Comment by Jognt** [comment]
Hi, I’m the poster of the mm-patch that implements skyrex’s temporary fix. 

The patch you’re talking about is to work around a Kopernicus 1.7.3-1 bug that impacts all installations of Kopernicus, both rescaled and stock scale. 

Besides that bug there is the observation that Kopernicus or its dependencies generate a LOT of garbage each frame. 

The graph you posted indicates hitting a CPU bottleneck. This may be due to the garbage creation, as it implies the use of an expensive bit of code. 

Whether it’s an outright bug or not I cannot say. Modding KSP is often subject to performance decreases, and it’s up to mod authors to decide if it’s worth spending time on to improve. 

The mm-patch should at the very least help with the massive FPS drop near/on surfaces when within unpack-distance of a craft. Which I think was the original issue you referred to. 

I highly suggest joining in the Kopernicus topic if you wish to troubleshoot some more since there are more people there that could help out. 

Hope this helps!

**Comment by Sigma88** [comment]
is this still happening with Kopernicus 1.7.3-2?

**Comment by Joker58th** [comment]
I just tested again with a fresh install and was unable to reproduce it.
